dialogy/plugins/text/calibration/xgb.py

```diff
@@ -0,0 +1,109 @@
+"""
+Trains a calibraation model. This contains two models:
+- Vectorizer: TfIdf
+- Classifier: XGBoostRegressor
+"""
+
+from _typeshed import OpenBinaryMode
+import json
+import math
+import pickle
+import sqlite3
+from typing import Any, Dict, List, Optional, Tuple
+
+import jiwer
+import numpy as np
+import pandas as pd
+import sklearn
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.pipeline import Pipeline
+from tqdm import tqdm
+from xgboost import XGBRegressor
+
+from dialogy.base.plugin import Plugin
+from dialogy.types import plugin
+
+
+class FeatureExtractor(BaseEstimator, TransformerMixin):
+    def __init__(self) -> None:
+        super().__init__()
+        self.vectorizer = TfidfVectorizer()
+
+    def read_data(self, fname: Any) -> pd.DataFrame:
+        if isinstance(fname, pd.DataFrame):
+            return fname
+        cux = sqlite3.connect(fname)
+        return pd.read_sql_query("SELECT * FROM DATA", cux)
+
+    def fit(self, filename: str, y=None):
+        df = self.read_sqlite(filename)
+        texts = []
+        for _, row in tqdm(df.iterrows()):
+            real_transcript = json.loads(row["tag"])["text"]
+            texts.append(real_transcript)
+            alts = json.loads(row["data"])["alternatives"]
+            if alts in [[], [None]]:
+                continue
+            for alt in alts[0]:
+                texts.append(alt["transcript"])
+
+        self.vectorizer.fit(texts)
+        return self
+
+    def features(self, alt: Dict) -> List:
+        l = len(alt["transcript"].split())
+        return self.vectorizer.transform([alt["transcript"]]).todense().tolist()[0] + [
+            alt["am_score"] / l,
+            alt["lm_score"] / l,
+            alt["transcript"].count("UNK"),
+            l,
+            alt["am_score"] / math.log(1 + l),
+            alt["lm_score"] / math.log(1 + l),
+            alt["am_score"] / math.sqrt(l),
+            alt["am_score"] / math.sqrt(l),
+        ]
+
+    def transform(self, filename: str) -> Tuple[np.ndarray, List]:
```

Comment:
Let's get the dataframe here. As an example, refer to [this line](https://github.com/skit-ai/dialogy/blob/9137e86197ec6f7d015b6c7c00945d7c11115e73/dialogy/plugins/text/merge_asr_output/__init__.py#L105). It is easier to provide small mock dataframes to plugins to test a few cases.